### PR TITLE
Convert SAPI to the MU API

### PR DESCRIPTION
**Pull request by tstruk**
Start conversion with CreatePrimary

### CODE REVIEW

File: sysapi/sysapi/EncryptParam.c [hunk 1]
```diff
@@ -93,9 +93,12 @@ TSS2_RC Tss2_Sys_SetEncryptParam(
...
             else
             {
-                CopyMem( (uint8_t *)currEncryptParamBuffer, encryptParamBuffer,
-                        encryptParamSize,
-                        SYS_CONTEXT->tpmInBuffPtr + SYS_CONTEXT->maxCommandSize );
+                if (currEncryptParamBuffer + encryptParamSize >
+                        SYS_CONTEXT->tpmInBuffPtr + SYS_CONTEXT->maxCommandSize)
+                    return TSS2_SYS_RC_INSUFFICIENT_CONTEXT;
+
+               memmove((UINT8 *)currEncryptParamBuffer,
```
Comment (flihp) [inline sysapi/sysapi/EncryptParam.c]: You're casting away the `const` here. If I'm remembering memmove right this is the pointer to the location where the data is moved (the destination) so I'm wondering if the const is appropriate at all?

File: sysapi/sysapi/Tss2_Sys_CreatePrimary.c [hunk 2]
```diff
@@ -30,117 +30,152 @@
...
-    {
-        rval = CommonOneCall( sysContext, cmdAuthsArray, rspAuthsArray );
+    rval = CommonOneCall(sysContext, cmdAuthsArray, rspAuthsArray);
 
-        if( rval == TSS2_RC_SUCCESS )
-        {
-            rval = Tss2_Sys_CreatePrimary_Complete( sysContext, objectHandle, outPublic, creationData, creationHash, creationTicket, name );
-        }
-    }
+    if (rval)
```
Comment (flihp) [inline sysapi/sysapi/Tss2_Sys_CreatePrimary.c]: Your whitespace is a bit inconsistent in these calls. I prefer to have no whitespace between the function call and the check on the return value.

File: sysapi/sysapi/Tss2_Sys_CreatePrimary.c [hunk 3]
```diff
@@ -30,117 +30,152 @@
...
+        return rval;
 
+    rval = Tss2_Sys_CreatePrimary_Complete(sysContext, objectHandle, outPublic,
+                                           creationData, creationHash,
+                                           creationTicket, name);
     return rval;
-}
 
+}
```
Comment (flihp) [inline sysapi/sysapi/Tss2_Sys_CreatePrimary.c]: Whitespace between return and the closing parenthesis departs from the norm.

File: sysapi/sysapi_util/unmarshal_simple_tpm2b.c [hunk 4]
```diff
@@ -29,67 +29,35 @@
...
+        return;
     }
-}
 
+    *rval = UINT16_Unmarshal(outBuffPtr, maxResponseSize, nextData, outTPM2B ? &outTPM2B->size : NULL);
+
+    if (*rval)
+        return;
+
+    for (i = 0; i < length; i++) {
```
Comment (flihp) [inline sysapi/sysapi_util/unmarshal_simple_tpm2b.c]: This may be best left to another commit but these for loops that do byte by byte copy operations have always seemed wasteful to me. What do you think about just using a memcopy here? Looks to me like we've already got the size of all relevant data (data size, destination buffer size etc) to do so safely.

File: sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c [hunk 5]
```diff
@@ -28,60 +28,30 @@
...
-                {
-                    // Let low level function deal with NULL output pointer.
-                    Unmarshal_UINT16( outBuffPtr, maxResponseSize, nextData, 0, rval );
-                }
+    if (!outBuffPtr || !nextData) {
+        *rval = TSS2_SYS_RC_BAD_REFERENCE;
+        return;
+    }
+
+    length = BE_TO_HOST_16(*(UINT16 *)(outBuffPtr + *nextData));
```
Comment (flihp) [inline sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c]: What do you think about using the appropriate unmarshalling function here? That would get us the missing boundary check here for free.

File: sysapi/sysapi_util/unmarshal_simple_tpm2b.c [hunk 6]
```diff
@@ -29,67 +29,35 @@
...
-                        if( outTPM2B != 0 )
-                        {
-                            Unmarshal_UINT8( outBuffPtr, maxResponseSize, nextData,  &( outTPM2B->buffer[i] ), rval );
-                        }
-                        else
-                        {
-                            // Let low level function deal with NULL output pointer.
-                            Unmarshal_UINT8( outBuffPtr, maxResponseSize, nextData, 0, rval );
-                        }
+    length = BE_TO_HOST_16(*(UINT16 *)(outBuffPtr + *nextData));
```
Comment (flihp) [inline sysapi/sysapi_util/unmarshal_simple_tpm2b.c]: What do you think about using the appropriate unmarshalling function here? That would get us the missing boundary check for free.

File: sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c [hunk 7]
```diff
@@ -28,60 +28,30 @@
...
 
-                        if( *rval != TSS2_RC_SUCCESS )
-                        {
-                            break;
-                        }
-                    }
-                }
-            }
-        }
+    for (i = 0; i < length; i++) {
```
Comment (flihp) [inline sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c]: Same comment as above w/r to memcpy

File: sysapi/sysapi/EncryptParam.c (hunk 1 above)
Comment (tstruk) [reply sysapi/sysapi/EncryptParam.c]: This is because the Tss2_Sys_GetEncryptParam() expects const uint8_t ** so we have to cast somewhere. I'm not sure which place is better to cast memmove or Tss2_Sys_GetEncryptParam.
For now I will change the declaration to uint8_t *currEncryptParamBuffer

File: sysapi/sysapi/Tss2_Sys_CreatePrimary.c (hunk 3 above)
Comment (tstruk) [reply sysapi/sysapi/Tss2_Sys_CreatePrimary.c]: Do you mean blank line?

File: sysapi/sysapi/Tss2_Sys_CreatePrimary.c (hunk 2 above)
Comment (tstruk) [reply sysapi/sysapi/Tss2_Sys_CreatePrimary.c]: Blank lines removed

File: sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c (hunk 5 above)
Comment (tstruk) [reply sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c]: We could, but I'm planning to remove all the tree (un)marshal_simple_tpm2b(_no_size_check).c
files when the conversion is done, and use the appropriate TPM2B_<TYPE>_Marshal/TPM2B_<TYPE>_Umarshal functions, so this is only temporary anyway.

File: sysapi/sysapi_util/unmarshal_simple_tpm2b.c (hunk 4 above)
Comment (tstruk) [reply sysapi/sysapi_util/unmarshal_simple_tpm2b.c]: Right, but as I said, this is only temporary and will be replaced the the proper TPM2B_TYPE_Marshal calls.

File: sysapi/sysapi/Tss2_Sys_CreatePrimary.c (hunk 3 above)
Comment (flihp) [reply sysapi/sysapi/Tss2_Sys_CreatePrimary.c]: yes

File: sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c (hunk 5 above)
Comment (flihp) [reply sysapi/sysapi_util/unmarshal_simple_tpm2b_no_size_check.c]: Understood. Comment withdrawn.

**Comment by tstruk** [comment]
this resolves #540
